Add best practices guide for pod auto scaling (#10083)
gardener/gardener: docs/usage/pod-auto-scaling-best-practices.md

> # Introduction

Reviewer: Nit: You could link the newly introduced document in the documentation index (`docs/README.md`).

Author: Excellent feedback. I see, I must do that and a bit more like updating the existing/outdated shoot autoscaling guide. Thanks for catching that! Will do.

Author: Done.

There are two types of pod autoscaling in Kubernetes: Horizontal Pod Autoscaling (HPA) and Vertical Pod Autoscaling (VPA). HPA (implemented as part of the kube-controller-manager) scales the number of pod replicas, while VPA (implemented as independent community project) adjusts the CPU and memory requests for the pods. Both types of autoscaling aim to optimize resource usage/costs and maintain the performance and (high) availability of applications running on Kubernetes.

> ## [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/de/docs/tasks/run-application/horizontal-pod-autoscale)

Reviewer: Nit: The current link points to the german (de) docs.

```suggestion
## [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
```

Author: Sorry, missed that when checking the links. Thanks. Done.

## [Vertical Pod Autoscaling (VPA)](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler#readme)

Vertical Pod Autoscaling, on the other hand, focuses on adjusting the CPU and memory resources allocated to the pods themselves. Instead of changing the number of replicas, VPA tweaks the resource requests (and limits, but only [proportionally](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/README.md#keeping-limit-proportional-to-request), if configured) for the pods in a deployment, replica set, stateful set, daemon set, or [anything really with a scale subresource](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/FAQ.md#how-can-i-apply-vpa-to-my-custom-resource). This means that each pod can be given more, or fewer resources as needed.

> VPA is very useful for optimizing the resource requests of pods that have dynamic resource needs over time. It does so by mutating pod requests (unfortunately, [not in-place](https://github.com/kubernetes/design-proposals-archive/blob/main/autoscaling/vertical-pod-autoscaler.md#in-place-updates)). Therefore, in order to apply new recommendations, pods that are "out of bounds" (i.e. below a configured/computed lower or above a configured/computed upper recommendation percentile) will be evicted proactively, but also pods that are "within bounds" may be evicted after a grace period. The [replication controller](https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller) will then recreate a new pod that VPA will then mutate to set the currently recommended requests (and proportional limits, if configured).

Reviewer: > The [replication controller](https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller) will then recreate [...]

`kind: ReplicationController` is a higher level controller that is replaced by ReplicaSet/Deployment. Additionally, StatefulSets, DaemonSets are not using ReplicaSet/ReplicationController under the hood. So, I would rather reword it like "the higher level controller will [...]"

Author: Yeah, I had something similar like "higher level" and "owner resource", but I didn't like it very much as it's so unspecific.

How about I drop the link, because that's clearly wrong and we keep the "general" word "replication controller"? Would that be OK (see updated PR)? If not, I will go with your proposal, of course.

Reviewer: That would be OK, let's drop the link only then.

Author: Thanks.

- **Scalability:** Can your application handle increased load by scaling vertically (more resources per pod) or does it require horizontal scaling (more pod instances)?

HPA is the right choice if:
- Your application is stateless and can handle increased load by adding more instances.
- You experience short-term fluctuations in traffic that require quick scaling responses.
- You want to maintain a specific performance metric, such as requests per second per pod.

VPA is the right choice if:
- Your application's resource requirements change over time, and you want to optimize resource usage without manual intervention.
- You want to avoid the complexity of managing resource requests for each pod, especially when they run code where it's impossible for you to suggest static requests.

In essence:
- For applications that can handle increased load by simply adding more replicas, HPA should be used to handle short-term fluctuations in load by scaling the number of replicas.
- For applications that require more resources per pod to handle additional work, VPA should be used to adjust the resource allocation for longer-term trends in resource usage.

> Consequently, if both cases apply (VPA often applies), HPA and VPA can also be combined. However, combining both, especially on the same metrics (CPU and memory), requires understanding and care to avoid conflicts and ensure that the autoscaling actions do not interfere with and rather complement each other.

Reviewer: ```suggestion
Consequently, if both cases apply (VPA often applies), HPA and VPA can also be combined. However, combining both, especially on the same metrics (CPU and memory), requires understanding and care to avoid conflicts and ensure that the autoscaling actions do not interfere with and rather complement each other. For more details, see [Combining HPA and VPA](#combining-hpa-and-vpa).
```

Author: Sure, done.

## Considerations When Using HPA

- **Selection of metrics:** Besides CPU and memory, HPA can also target custom or external metrics. Pick those (in addition or exclusively), if you guarantee certain SLOs in your SLAs.
- **Targeting usage or utilization:** HPA supports usage (absolute) and utilization (relative). Utilization is often preferred in simple examples, but usage is more precise and versatile.
- **Compatibility with VPA:** Care must be taken when using HPA in conjunction with VPA, as they can potentially interfere with each other's scaling decisions.

# Vertical Pod Autoscaler (VPA)

> VPA operates by monitoring resource metrics for all pods in a target. It computes a resource requests recommendation from the historic and current resource metrics. VPA checks the metrics at regular intervals, which can be configured by the user. Only CPU and memory are supported. If VPA detects that a pod's resource allocation is too high or too low, it may evict pods (if within the permitted disruption budget), which will trigger the creation of a new pod by the corresponding [replication controller](https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller), which will then be mutated by VPA to match resource requests recommendation. This happens in three different components that work together:

Reviewer: ```suggestion
VPA operates by monitoring resource metrics for all pods in a target. It computes a resource requests recommendation from the historic and current resource metrics. VPA checks the metrics at regular intervals, which can be configured by the user. Only CPU and memory are supported. If VPA detects that a pod's resource allocation is too high or too low, it may evict pods (if within the permitted disruption budget), which will trigger the creation of a new pod by the corresponding higher level controller, which will then be mutated by VPA to match resource requests recommendation. This happens in three different components that work together:
```

Author: See above, let's keep the discussion in one place (first occurrence).

There are many more options that you can only configure if you deploy your own CA and which we will not discuss here, but you can check them out [here](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-are-the-parameters-to-ca).

# Importance of Monitoring

Monitoring is a critical component of autoscaling for several reasons:
- **Performance Insights:** It provides insights into how well your autoscaling strategy is meeting the performance requirements of your applications.
- **Resource Utilization:** It helps you understand resource utilization patterns, enabling you to optimize resource allocation and reduce waste.
- **Cost Management:** It allows you to track the cost implications of scaling actions, helping you to maintain control over your cloud spending.
- **Troubleshooting:** It enables you to quickly identify and address issues with autoscaling, such as unexpected scaling behavior or resource bottlenecks.

To effectively monitor autoscaling, you should leverage the following tools and metrics:

> - **Kubernetes Metrics Server:** Collects resource metrics from kubelets and provides them to HPA and VPA for autoscaling decisions.

Reviewer: We could add a note that a metrics-server is already provided for Gardener managed Shoot clusters.

Author: Sure, done, thanks.

- **Pod Count:** Monitor the number of pod replicas over time to see how HPA is responding to changes in load.
- **Scaling Events:** Keep an eye on scaling events triggered by HPA and VPA to ensure they align with expected behavior.
- **Application Performance Metrics:** Track application-specific metrics such as response times, error rates, and throughput.

Based on the insights gained from monitoring, you may need to adjust your autoscaling configurations:
- **Refine Thresholds:** If you notice frequent scaling actions or periods of underutilization or overutilization, adjust the thresholds used by HPA and VPA to better match the workload patterns.
- **Update Policies:** Modify VPA update policies if you observe that the current settings are causing too much or too little pod disruption.
- **Custom Metrics:** If using custom metrics, ensure they accurately reflect the load on your application and adjust them if they do not.
- **Scaling Limits:** Review and adjust the minimum and maximum scaling limits to prevent over-scaling or under-scaling based on the capacity of your cluster and the criticality of your applications.

> Best practices:

Reviewer: Shouldn't this be a separate section for best practices instead of having it included in the `Importance of Monitoring` section?
I see that the `Summary` section also defines best practices - we could merge both sections or more the bullets from here there.

Author: Absolutely, I changed the docs a few times and that didn't make it. I will move it down/de-duplicate it.

Author: Done.

- If your pods have multiple containers, avoid scaling vertically on all of them. Most of the time there is only one main container that requires vertical scaling while the other containers are relatively static sidecars that work fine with static requests.
- If you have containers with static requests, it is a good rule of thumb to set the 95th or 90th pod usage percentile as static requests.

> - If you have containers under VPA, it is a good rule of thumb to set the 5th or 10th pod usage percentile as initial requests.

Reviewer: Resource requests are per container, I guess we rather mean:
```suggestion
- If you have containers under VPA, it is a good rule of thumb to set the 5th or 10th container usage percentile as initial requests.
```

Author: Good catch, indeed, copy&paste error. Thank you, corrected.

- If you have defined absolute limits (unrelated to the requests), configure VPA to only scale the requests or else it will proportionally scale the limits as well, which can easily become useless (way beyond the threshold of unhealthy behavior) or absurd (larger than node capacity):
  ```
  spec:
    resourcePolicy:
      containerPolicies:
      - controlledValues: RequestsOnly
        ...
  ```
  If you have defined relative limits (related to the requests), the default policy to scale the limits proportionally with the requests is fine, but the gap between requests and limits must be zero for QoS `Guaranteed` and should best be small for QoS `Burstable` to avoid useless or absurd limits either, e.g. prefer limits being 5 to at most 20% larger than requests as opposed to being 100% larger or more.

> - As a rule of thumb, set minAllowed to the highest observed VPA recommendation (usually during the initialization phase or during any periodical activity) for an otherwise practically idle container, so that you avoid needless trashing (e.g. resource usage calms down over time and recommendations drop consecutively until eviction, which will then lead again to initialization or later periodical activity and higher recommendations and new evictions).<br/>⚠️ You may want to provide higher minAllowed values, if you observe that up-scaling takes too long for CPU or memory for a too large percentile of your workload. This will get you out of the danger zone of too few resources for too many pods at the expense of providing too many resources for a few pods. Memory may react faster than CPU, because CPU throttling is not visible and memory gets aided by OOM bump-up incidents, but still, if you observe that up-scaling takes too long, you may want to increase minAllowed accordingly.

Reviewer: As another best practice, we could add is "when using K8s operator that deploys workload and HPA for it, do not overwrite the higher level controller's replicas". We had such bugs in the past where gardenlet or GRM was overwriting the replicas count of a higher level controller during reconciliation. This results in a revert of the value set/managed by HPA. Hence, the operator should not overwrite the higher level controller's replicas count when it is scaled by HPA.

Author: Indeed, let me add that, thank you. Maybe even a word on VPA and the vpacheckpoints and the consequences if the deployments (etc.) get updated and the web hook is down? Then again, that would be on us.

Author: I updated:
- As a rule of thumb, set the initial replicas to the 5th percentile of the actually observed replica count in production. Since HPA reacts fast, this is not as crucial, but may help reduce initial load on the control plane early after deployment. However, be cautious when you update the higher-level resource not to inadvertently reset the current HPA-controlled replica count (very easy to make mistake that can lead to catastrophic loss of pods). HPA modifies the replica count directly in the spec and you do not want to overwrite that. Even if it reacts fast, it is not instant (not via a mutating webhook as VPA operates) and the damage may already be done.

And:
- As a rule of thumb, set the initial requests to the 5th percentile of the actually observed CPU resp. memory usage in production. Since VPA may need some time at first to respond and evict pods, this is especially critical early after deployment. The lower bound, below which pods will be immediately evicted, converges much faster than the upper bound, above which pods will be immediately evicted, but it isn't instant, e.g. after 5 minutes the lower bound is just at 60% of the computed lower bound; after 12 hours the upper bound is still at 300% of the computed upper bound (see [code](https://github.com/kubernetes/autoscaler/blob/b3a501cbe11e46bea1f8879d39c8436ef03e7139/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L118-L143)). Unlike with HPA, you don't need to be as cautious when updating the higher-level resource in the case of VPA. As long as VPA's mutating webhook (VPA Admission Controller) is operational (which also the VPA Updater checks before evicting pods), it's generally safe to update the higher-level resource. However, if you want to err on the side of caution, it's advisable to check the health of the VPA Admission Controller first. If it's not up and running, any new pods that are spawned (e.g. as a consequence of a rolling update of the higher-level resource or for any other reason) will not be mutated. Instead, they will receive whatever requests are currently configured at the higher-level resource, which can lead to catastrophic resource under-reservation. Gardener deploys the VPA Admission Controller in HA - if unhealthy, it is reported under the `ControlPlaneHealthy` shoot status condition.

Author: Is the passage on VPA OK or should I drop that sentence/most of the paragraph, assuming the Admission Controller to be always healthy (our job)?
> Gardener deploys the VPA Admission Controller in HA - if unhealthy, it is reported under the ControlPlaneHealthy shoot status condition.

Reviewer 2: Hmm the status might not be updated immediately, right? So even with that users could run into issues. Maybe just giving them a fair warning that this could occur so that they have it in mind when modifying the corresponding higher-level resources is enough.

Author: Thanks also @plkokanov. I removed the "err on the side of caution" sentence. Now:
> As a rule of thumb, set the initial requests to the 5th percentile of the actually observed CPU resp. memory usage in production. Since VPA may need some time at first to respond and evict pods, this is especially critical early after deployment. The lower bound, below which pods will be immediately evicted, converges much faster than the upper bound, above which pods will be immediately evicted, but it isn't instant, e.g. after 5 minutes the lower bound is just at 60% of the computed lower bound; after 12 hours the upper bound is still at 300% of the computed upper bound (see [code](https://github.com/kubernetes/autoscaler/blob/b3a501cbe11e46bea1f8879d39c8436ef03e7139/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go#L118-L143)). Unlike with HPA, you don't need to be as cautious when updating the higher-level resource in the case of VPA. As long as VPA's mutating webhook (VPA Admission Controller) is operational (which also the VPA Updater checks before evicting pods), it's generally safe to update the higher-level resource. However, if it's not up and running, any new pods that are spawned (e.g. as a consequence of a rolling update of the higher-level resource or for any other reason) will not be mutated. Instead, they will receive whatever requests are currently configured at the higher-level resource, which can lead to catastrophic resource under-reservation. Gardener deploys the VPA Admission Controller in HA - if unhealthy, it is reported under the `ControlPlaneHealthy` shoot status condition.
